README.md
# Drone Conversion Extension: Paths Changed | [![Build Status](https://cloud.drone.io/api/badges/meltwater/drone-convert-pathschanged/status.svg)](https://cloud.drone.io/meltwater/drone-convert-pathschanged) [![Docker Pulls](https://img.shields.io/docker/pulls/meltwater/drone-convert-pathschanged)](https://hub.docker.com/r/meltwater/drone-convert-pathschanged)

A [Drone](https://drone.io/) [conversion extension](https://docs.drone.io/extensions/conversion/) to include/exclude pipelines and steps based on paths changed.

_Please note this project requires Drone server version 1.4 or higher._

## Installation

## Github Cloud

1. Create a github token via https://github.com/settings/tokens with the scope of`repo` (see [issue 13](https://github.com/meltwater/drone-convert-pathschanged/issues/13) for background).

2. Create a shared secret:

```console
$ openssl rand -hex 16
bea26a2221fd8090ea38720fc445eca6
```

3. Download and run the plugin:

```console
$ docker run -d \
  --publish=3000:3000 \
  --env=DRONE_DEBUG=true \
  --env=DRONE_SECRET=bea26a2221fd8090ea38720fc445eca6 \
  --env=TOKEN=9e6eij3ckzvpe9mrhnqcis6zf8dhopmm46e3pi96 \
  --env=PROVIDER=github \
  --restart=always \
  --name=converter meltwater/drone-convert-pathschanged
```
If you wish to use an enviroment file you can pass it when starting the container :
```console
$ docker run -d \
...
  --name=converter meltwater/drone-convert-pathschanged --envfile drone.env
```

4. Update your Drone server configuration to include the plugin address and the shared secret.

```text
DRONE_CONVERT_PLUGIN_ENDPOINT=http://1.2.3.4:3000
DRONE_CONVERT_PLUGIN_SECRET=bea26a2221fd8090ea38720fc445eca6
```

## Github Server

1. Create a github token via https://your-github-server-address/settings/token with the scope of `repo`

2. Create a shares secret:

```console
$ openssl rand -hex 16
bea26a2221fd8090ea38720fc445eca6
```

3. Download ran run the plugin:

```console
$ docker run -d \
  --publish=3000:3000 \
  --env=DRONE_DEBUG=true \
  --env=DRONE_SECRET=bea26a2221fd8090ea38720fc445eca6 \
  --env=TOKEN=9e6eij3ckzvpe9mrhnqcis6zf8dhopmm46e3pi96 \
  --env=PROVIDER=github \
  --env=GITHUB_SERVER=https://your-github-server-address
  --restart=always \
  --name=converter meltwater/drone-convert-pathschanged
```

4. Update your Drone server configuration to include the plugin address and the shared secret.

```text
DRONE_CONVERT_PLUGIN_ENDPOINT=http://1.2.3.4:3000
DRONE_CONVERT_PLUGIN_SECRET=bea26a2221fd8090ea38720fc445eca6
```

## Bitbucket Cloud

1.  Create an "App password" via https://bitbucket.org/account/settings/app-passwords and select only "Read" under "Repositories"

2. Create a shared secret:

```console
$ openssl rand -hex 16
bea26a2221fd8090ea38720fc445eca6
```

3. Download and run the plugin:

```console
$ docker run -d \
  --publish=3000:3000 \
  --env=DRONE_DEBUG=true \
  --env=DRONE_SECRET=bea26a2221fd8090ea38720fc445eca6 \
  --env=BITBUCKET_USER=youruser \
  --env=BITBUCKET_PASSWORD='yourpassword' \
  --env=PROVIDER=bitbucket \
  --restart=always \
  --name=converter meltwater/drone-convert-pathschanged
```

4. Update your Drone server configuration to include the plugin address and the shared secret.

```text
DRONE_CONVERT_PLUGIN_ENDPOINT=http://1.2.3.4:3000
DRONE_CONVERT_PLUGIN_SECRET=bea26a2221fd8090ea38720fc445eca6
```

## Stash (Bitbucket Server)

1. Create a Stash access token via https://your-bitbucket-address/plugins/servlet/access-tokens/manage with read-only rights

2. Create a shared secret:

```console
$ openssl rand -hex 16
bea26a2221fd8090ea38720fc445eca6
```

3. Download and run the plugin:

```console
$ docker run -d \
  --publish=3000:3000 \
  --env=DRONE_DEBUG=true \
  --env=DRONE_SECRET=bea26a2221fd8090ea38720fc445eca6 \
  --env=TOKEN=9e6eij3ckzvpe9mrhnqcis6zf8dhopmm46e3pi96 \
  --env=PROVIDER=bitbucket-server \
  --env=STASH_SERVER=https://your-bitbucket-server-address
  --restart=always \
  --name=converter meltwater/drone-convert-pathschanged
```
Stash APIs are paginated and by default will only return 25 changes when computing the changeset. Set environment variable STASH_PAGE_SIZE to a higher number on the plugin to avoid partial change set detection.
```console
--env=STASH_PAGE_SIZE=1000
```

4. Update your Drone server configuration to include the plugin address and the shared secret.

```text
DRONE_CONVERT_PLUGIN_ENDPOINT=http://1.2.3.4:3000
DRONE_CONVERT_PLUGIN_SECRET=bea26a2221fd8090ea38720fc445eca6
```

## Gitee Cloud

1. Create a gitee token via https://gitee.com/personal_access_tokens with the scope of`repo` (see [issue 13](https://github.com/meltwater/drone-convert-pathschanged/issues/13) for background).

2. Create a shared secret:

```console
$ openssl rand -hex 16
bea26a2221fd8090ea38720fc445eca6
```

3. Download and run the plugin:

```console
$   docker run -d \
  --publish=3000:3000 \
  --env=DRONE_DEBUG=true \
  --env=DRONE_SECRET=bea26a2221fd8090ea38720fc445eca6 \
  --env=TOKEN=b4af6fc778a7aba5f2a133d155f5b4a9cbe7becb255557e67597a4967eb50a88 \
  --env=PROVIDER=gitee \
  --restart=always \
  --name=converter meltwater/drone-convert-pathschanged
```

4. Update your Drone server configuration to include the plugin address and the shared secret.

```text
DRONE_CONVERT_PLUGIN_ENDPOINT=http://1.2.3.4:3000
DRONE_CONVERT_PLUGIN_SECRET=bea26a2221fd8090ea38720fc445eca6
```

## Examples

This extension uses [doublestar](https://github.com/bmatcuk/doublestar) for matching paths changed in your commit range, refer to their documentation for all supported patterns.

### `include`

Only run a pipeline when `README.md` is changed:
```yaml
---
kind: pipeline
name: readme

trigger:
  paths:
    include:
    - README.md

steps:
- name: message
  image: busybox
  commands:
  - echo "README.md was changed”
```

Only run a pipeline step when `README.md` is changed:
```yaml
---
kind: pipeline
name: readme

steps:
- name: message
  image: busybox
  commands:
  - echo "README.md was changed”
  when:
    paths:
      include:
      - README.md
```

Same as above, but with an implicit `include`:
```yaml
---
kind: pipeline
name: readme

steps:
- name: message
  image: busybox
  commands:
  - echo "README.md was changed”
  when:
    paths:
    - README.md
```

### `include` and `exclude`

Run a pipeline step when `.yml` files are changed in the root, except for `.drone.yml`:
```yaml
---
kind: pipeline
name: yaml

steps:
- name: message
  image: busybox
  commands:
  - echo "A .yml file in the root of the repo other than .drone.yml was changed"
  when:
    paths:
      include:
      - "*.yml"
      exclude:
      - .drone.yml
```

### `depends_on`

When using [`depends_on`](https://docker-runner.docs.drone.io/configuration/parallelism/) in a pipeline step, ensure the `paths` rules match, otherwise your steps may run out of order.

Only run two steps when `README.md` is changed, one after the other:
```yaml
---
kind: pipeline
name: depends_on

steps:
- name: message
  image: busybox
  commands:
  - echo "README.md was changed”
  when:
    paths:
      include:
      - README.md

- name: depends_on_message
  depends_on:
  - message
  image: busybox
  commands:
  - echo "This step runs after the message step"
  when:
    paths:
      include:
      - README.md
```

## Changesets

The changeset is generated by comparing the list of files changed between the commit [before] the patch is applied and the commit [after] the patch is applied.
As a result, the changeset for a commit may be different depending on which type of event triggered the build.

For example, the `push` and `tag` events may generate a changeset against the previous commit, where as the `pull_request` event may generate a changeset against the source branch. For more specifics on how before and after are set, review the webhook [parser].

[before]: https://docs.drone.io/pipeline/environment/reference/drone-commit-before/
[after]: https://docs.drone.io/pipeline/environment/reference/drone-commit-after/
[parser]: https://github.com/drone/drone/blob/ca454594021099909fb4ee9471720cacfe3207bd/service/hook/parser/parse.go

## Known issues

### Empty commits

Be careful when making empty commits with [`git commit --allow-empty`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---allow-empty). When an empty commit is made, no files have changed, so this plugin will return the unmodified `.drone.yml` back to the drone server process.

This can lead to potentially unexpected behavior, since any `include` or `exclude` rules will effectively be ignored.

### YAML anchors

There is a problem in the YAML library where ordering matters during unmarshaling, see https://github.com/meltwater/drone-convert-pathschanged/issues/18

This syntax will fail:

```yaml
anchor: &anchor
  image: busybox
  settings:
    foo: bar

- name: test
  <<: *anchor
  when:
    event: push
    branch: master
```

But this will succeed:

```yaml
anchor: &anchor
  image: busybox
  settings:
    foo: bar

- <<: *anchor 
  name: test
  when:
    event: push
    branch: master
```

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) to understand how to submit pull requests to us, and also see our [code of conduct](CODE_OF_CONDUCT.md).

### Protected Repos

When this plugin is used in conjunction with [protected repos](https://docs.drone.io/signature/),
signature validation will frequently fail.

This occurs due to Drone's order of operations, in that the Drone file's 
signature is checked after the this plugin has rewritten sections based on 
the paths-changed triggers, resulting in a different signature for the file.


CHANGELOG.md
## Unreleased

### Added
- [#89](https://github.com/meltwater/drone-convert-pathschanged/issues/89) add environment variable to allow stash users to specify page size of git compare changes

## 1.0.0

### Breaking changes
- [#79](https://github.com/meltwater/drone-convert-pathschanged/pull/79) shift to drone/go-scm library support

### Added
- [#83](https://github.com/meltwater/drone-convert-pathschanged/pull/83) add missing environnment variable to stash (bitbucket-server) example
- [#77](https://github.com/meltwater/drone-convert-pathschanged/pull/77) add bitbucket cloud support
- [#76](https://github.com/meltwater/drone-convert-pathschanged/pull/76) golang module updates
- [#74](https://github.com/meltwater/drone-convert-pathschanged/pull/74) update drone/go and drone/go-scm modules
- [#73](https://github.com/meltwater/drone-convert-pathschanged/pull/74) golang tidy

## 0.6.0
### Added
- [#72](https://github.com/meltwater/drone-convert-pathschanged/pull/72) add GitHub Server support

## 0.5.0
### Breaking changes
- [#47](https://github.com/meltwater/drone-convert-pathschanged/pull/47) build images from 'scratch'
  - Official documentation https://docs.docker.com/develop/develop-images/baseimages/#create-a-simple-parent-image-using-scratch

## 0.4.0
### Breaking changes
- [#46](https://github.com/meltwater/drone-convert-pathschanged/pull/46) handle `--allow-empty` commits with excludes
  - Until this fix, `--allow-empty` commits with pipeline or path exclusion rules would have caused the pipeline or path to be excluded. With this fix, exclusions are handled properly, so empty commits will trigger pipelines and pipeline steps with path exclusion rules (since no files were changed, pipelines and steps should not be excluded). This could be potentially unexpected behavior if you had been relying on the exclusion of pipelines or steps when making empty commits.

## 0.3.1
### Added
- linux-arm64 build

## 0.3.0
### Added
- [#38](https://github.com/meltwater/drone-convert-pathschanged/pull/38) Make include implicit and is optional
    - This addresses [#33](https://github.com/meltwater/drone-convert-pathschanged/issues/33) to make this plugin compatble with [`drone jsonnet`](https://docs.drone.io/pipeline/scripting/jsonnet/)
- [#40](https://github.com/meltwater/drone-convert-pathschanged/pull/40) Additional tests

## 0.2.0
### Added
- [#29](https://github.com/meltwater/drone-convert-pathschanged/pull/29) Initial experimental Bitbucket Server support.
    - This includes breaking changes, environment variable `GITHUB_TOKEN` has been replaced with `TOKEN`, and `PROVIDER` is a new required environment variable which must be either `github` or `bitbucket-server`.

## 0.1.0
### Added
- Initial prometheus support


main.go
// Copyright 2019 the Drone Authors. All rights reserved.
// Use of this source code is governed by the Blue Oak Model License
// that can be found in the LICENSE file.

package main

import (
	"flag"
	"fmt"
	"io"
	"net/http"

	"github.com/meltwater/drone-convert-pathschanged/plugin"

	"github.com/drone/drone-go/plugin/converter"
	"github.com/joho/godotenv"
	_ "github.com/joho/godotenv/autoload"
	"github.com/kelseyhightower/envconfig"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/sirupsen/logrus"
)

// spec provides the plugin settings.
type (
	spec struct {
		Bind   string `envconfig:"DRONE_BIND"`
		Debug  bool   `envconfig:"DRONE_DEBUG"`
		Text   bool   `envconfig:"DRONE_LOGS_TEXT"`
		Secret string `envconfig:"DRONE_SECRET"`

		Provider string `envconfig:"PROVIDER"`
		Token    string `envconfig:"TOKEN"`
		// BB_ADDRESS is deprecated in favor of STASH_SERVER, it will be removed in a future version
		BitBucketAddress  string `envconfig:"BB_ADDRESS"`
		BitBucketUser     string `envconfig:"BITBUCKET_USER"`
		BitBucketPassword string `envconfig:"BITBUCKET_PASSWORD"`
		GithubServer      string `envconfig:"GITHUB_SERVER"`
		StashServer       string `envconfig:"STASH_SERVER"`
		StashPageSize     int    `envconfig:"STASH_PAGE_SIZE"` //bumps the rest api page size for changeset detection. Picks only 25 changes if unspecified
	}
)

func contains(s []string, str string) bool {
	for _, v := range s {
		if v == str {
			return true
		}
	}
	return false
}

func validate(spec *spec) error {
	if spec.Secret == "" {
		return fmt.Errorf("missing secret key")
	}
	if spec.Provider == "" {
		return fmt.Errorf("missing provider")
	} else {
		providers := []string{
			"bitbucket",
			// bitbucket-server support is deprecated in favor of stash, it will be removed in a future version
			"bitbucket-server",
			"github",
			"stash",
			"gitee",
		}
		if !contains(providers, spec.Provider) {
			return fmt.Errorf("unsupported provider")
		}
	}
	if spec.Token == "" && (spec.Provider == "github" || spec.Provider == "bitbucket-server" || spec.Provider == "stash") {
		return fmt.Errorf("missing token")
	}
	if spec.BitBucketUser == "" && spec.Provider == "bitbucket" {
		return fmt.Errorf("missing bitbucket user")
	}
	if spec.BitBucketPassword == "" && spec.Provider == "bitbucket" {
		return fmt.Errorf("missing bitbucket password")
	}
	if spec.BitBucketAddress == "" && spec.Provider == "bitbucket-server" {
		return fmt.Errorf("missing bitbucket server address")
	} else if spec.BitBucketAddress != "" && spec.Provider == "bitbucket-server" {
		// backwards compatible support for bitbucket-server, this will be removed in a future version
		spec.StashServer = spec.BitBucketAddress
		spec.Provider = "stash"

		logrus.Warningln("bitbucket-server support is deprecated, please use stash")
	}
	if spec.StashServer == "" && spec.Provider == "stash" {
		return fmt.Errorf("missing stash server")
	}

	if spec.Token == "" && spec.Provider == "gitee" {
		return fmt.Errorf("missing gitee token")
	}

	return nil
}

func main() {
	envfilepath := flag.String("envfile", "", "pass filepath to env file (optional)")
	flag.Parse()
	if *envfilepath != "" {
		godotenv.Load(*envfilepath)
	}
	spec := new(spec)
	err := envconfig.Process("", spec)
	if err != nil {
		logrus.Fatalln(err)
	}

	if spec.Debug {
		logrus.SetLevel(logrus.DebugLevel)
	}
	if spec.Text {
		logrus.SetFormatter(&logrus.TextFormatter{})
	} else {
		logrus.SetFormatter(&logrus.JSONFormatter{})
	}

	err = validate(spec)
	if err != nil {
		logrus.Fatalln(err)
	}

	if spec.Bind == "" {
		spec.Bind = ":3000"
	}

	params := &plugin.Params{
		BitBucketUser:     spec.BitBucketUser,
		BitBucketPassword: spec.BitBucketPassword,
		GithubServer:      spec.GithubServer,
		Token:             spec.Token,
		StashServer:       spec.StashServer,
		StashPageSize:     spec.StashPageSize,
	}

	handler := converter.Handler(
		plugin.New(spec.Provider, params),
		spec.Secret,
		logrus.StandardLogger(),
	)

	logrus.Infof("server listening on address %s", spec.Bind)

	http.Handle("/", handler)
	http.HandleFunc("/healthz", healthz)
	http.Handle("/metrics", promhttp.Handler())
	logrus.Fatal(http.ListenAndServe(spec.Bind, nil))
}

func healthz(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	io.WriteString(w, "OK")
}
